### PR TITLE
feat: add agent skills and use_agent to /strands command

### PR DESCRIPTION
**Pull request by agent-of-mkmeral**
## Summary

Adds a **beta agent** with its own runner to the `/strands` command. The beta agent has extended capabilities (skills, sub-agents) while the standard agent remains completely unchanged.

### Design: Same pipeline, different agent

```
/strands <command>        → standard agent (agent_runner.py — UNCHANGED)
/strands beta <command>   → beta agent (beta_agent_runner.py — NEW)
```

Both go through the same pipeline: **parse → run → finalize**. The only difference is which runner executes.

### Changes to Existing Files

| File | Change |
|------|--------|
| `agent_runner.py` | **ZERO changes** — identical to main |
| `process-input.cjs` | Detect `/strands beta ...`, set `agent_type` flag, route modes |
| `action.yml` | Read `agent_type`, conditionally run the right runner + skills setup |

### New Files

| File | Purpose |
|------|---------|
| `beta_agent_runner.py` | Beta agent runner — imports shared infra from `agent_runner.py`, adds AgentSkills plugin + `use_agent` tool |
| `agent-skills/task-adversarial-tester/SKILL.md` | Adversarial testing skill |
| `agent-skills/task-release-digest/SKILL.md` | Release digest skill with sub-agent orchestration |

### New Commands

| Command | Description |
|---------|-------------|
| `/strands beta` | Run the beta agent (default mode based on context) |
| `/strands beta adversarial-test` | Break PRs with edge case tests, security holes, and failure modes |
| `/strands beta release-digest` | Generate release digests across packages using sub-agents |
| `/strands beta implement` | Implement with beta agent (skills + use_agent available) |
| `/strands beta review` | Review with beta agent |

Existing commands (`/strands`, `/strands implement`, `/strands review`, `/strands refine`, `/strands release-notes`) are **completely unchanged**.

### How the beta agent differs from standard

| Aspect | Standard | Beta |
|--------|----------|------|
| Runner | `agent_runner.py` | `beta_agent_runner.py` |
| Tools | Standard tools | Standard tools + `use_agent` |
| Plugins | None | AgentSkills (SKILL.md format) |
| Skills dir | Not copied | Copied to working dir |
| Pipeline | parse → run → finalize | parse → run → finalize (same) |
| Impact on each other | None | None |

### Architecture

`beta_agent_runner.py` imports shared infrastructure from `agent_runner.py`:
- `_get_all_tools()` — base tools (beta adds on top)
- `_setup_langfuse_telemetry()` — telemetry setup
- `_get_trace_attributes()` — trace attributes
- `_send_eval_trigger()` — eval triggers
- Model constants (`STRANDS_MODEL_ID`, etc.)

This means the beta agent automatically inherits any improvements to the shared infra without duplication.

---

Closes the intent from #43. cc @mkmeral

### CODE REVIEW

**Comment by mkmeral** [comment]
I don't want to set up separate workflows. it should be under the same strands command so it works in all repos where we have strands command already

**Comment by mkmeral** [comment]
@agent-of-mkmeral ☝️

**Comment by agent-of-mkmeral** [comment]
Reworked per your feedback — everything is now integrated into the existing `/strands` command. No separate workflows or actions.

### What changed

**Before (6 new files, 0 modifications):**
- Separate `/strands-beta` command, parser, runner, and actions
- Repos would need to set up new workflows

**After (3 modifications, 2 new files):**
- `process-input.cjs` — added `adversarial-test` and `release-digest` mode routing
- `agent_runner.py` — added `use_agent` tool + `AgentSkills` plugin (graceful degradation)
- `action.yml` — checkout and copy `agent-skills/` to working directory
- `agent-skills/task-adversarial-tester/SKILL.md` (new)
- `agent-skills/task-release-digest/SKILL.md` (new)

### How it works

Any repo with `/strands` already set up gets these automatically:
- `/strands adversarial-test` — adversarial testing via AgentSkills plugin
- `/strands release-digest` — release digest with `use_agent` sub-agents
- All existing commands unchanged

If the `AgentSkills` plugin or `agent-skills/` directory aren't available, the agent works exactly as before — zero risk to existing behavior.

**Comment by mkmeral** [comment]
no, so i want like a separate agent with it's own runner, sowe can add more tools, plugins to it. but our changes should have no impact to the normal agent. 

that said, i want it to use the same pipeline overall, so don't recreate everything. like make agent creation configurable or something. so `/strands beta ...` is going to through the same pipeline, just to a different agent @agent-of-mkmeral

**Comment by agent-of-mkmeral** [comment]
Reworked per your feedback @mkmeral — the beta agent now has its own runner, but uses the same pipeline.

### What changed in this commit

**`agent_runner.py` — ZERO changes** (reverted to main, identical). Normal agent is completely untouched.

**`beta_agent_runner.py` — NEW** — Separate runner that:
- Imports shared infra from `agent_runner.py` (telemetry, trace attributes, tools, model config, eval triggers)
- Adds AgentSkills plugin + `use_agent` tool on top
- Easy to extend independently — add whatever tools/plugins you want here without touching the normal agent

**`process-input.cjs`** — Detects `/strands beta ...`, strips the prefix, sets `agent_type: "beta"` in the parsed output. Everything else routes normally.

**`action.yml`** — Reads `agent_type` and conditionally:
- Copies `agent-skills/` to working dir (beta only)
- Runs `beta_agent_runner.py` instead of `agent_runner.py` (beta only)

### The flow

```
/strands review         → process-input → agent_runner.py      (standard)
/strands beta review    → process-input → beta_agent_runner.py (beta)
```

Same pipeline (parse → run → finalize). Different agent.

### Diff vs main

```
 action.yml              |  30 ++-   (conditional runner + skills copy)
 task-adversarial-tester  | 108 +++   (new skill)
 task-release-digest      | 107 +++   (new skill)
 process-input.cjs        |  67 ++-   (beta routing)
 beta_agent_runner.py     | 209 +++++ (new runner)
 agent_runner.py          |   0       (UNCHANGED)
```

5 files changed, 511 insertions, 10 deletions. Zero modifications to `agent_runner.py`.

File: strands-command/scripts/javascript/process-input.cjs [hunk 1]
```diff
@@ -76,11 +76,39 @@ async function determineBranch(github, context, issueId, mode, isPullRequest) {
...
+  if (agentType === 'beta') {
+    const skillNameMap = {
+      'adversarial-test': 'task-adversarial-tester',
+      'release-digest': 'task-release-digest',
+    };
+    const skillName = skillNameMap[mode];
+
+    let systemPrompt;
+    if (skillName) {
+      systemPrompt = `You are an autonomous GitHub agent powered by Strands Agents SDK.
```
Comment (mkmeral) [inline strands-command/scripts/javascript/process-input.cjs]: I mean we should probably have a meaningful system prompt right? check your own system prompt, check the existing SOPs, [agent guidelines](https://github.com/strands-agents/docs/blob/main/team/AGENT_GUIDELINES.md) and so on. Let's have a proper system prompt

File: strands-command/scripts/python/beta_agent_runner.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+    Starts with all standard tools, then adds beta-only tools.
+    This ensures the beta agent is a strict superset of the standard agent.
+    """
+    tools = _get_all_tools()
+
+    # Add beta-only tools (use_agent is already imported at module level)
+    # Check if use_agent is already in the list (it is in the current version)
+    tool_names = {getattr(t, '__name__', str(t)) for t in tools}
+    if 'use_agent' not in tool_names:
```
Comment (mkmeral) [inline strands-command/scripts/python/beta_agent_runner.py]: let's also add https://github.com/strands-agents/tools/pull/387 to here, you can just create the tool here for now

File: strands-command/scripts/python/beta_agent_runner.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+                prefix=os.getenv("GITHUB_REPOSITORY", ""),
+            )
+        else:
+            raise ValueError("Both SESSION_ID and S3_SESSION_BUCKET must be set")
+
+        # Beta-only: Load agent skills plugin
+        plugins = []
+        skills_plugin = _load_skills_plugin()
+        if skills_plugin:
+            plugins.append(skills_plugin)
```
Comment (mkmeral) [inline strands-command/scripts/python/beta_agent_runner.py:362]: can we also activate a certain skill already based on the next word after beta? similar to SOPs,we can just invoke a skill like  `agent.tool.skills("activate", "reviewer")` (check skills tool description to make sure lol, i just made the signature up)

File: strands-command/agent-skills/task-adversarial-tester/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
+---
+name: task-adversarial-tester
```
Comment (mkmeral) [inline strands-command/agent-skills/task-adversarial-tester/SKILL.md:2]: add the meta reasoning skill too

also can we use the currently existing SOPs as skill too (without duplicating the files?) maybe we can just copy/paste on runtime, rename the files/folders and add frontmatter?

**Review by mkmeral — COMMENTED** [review]
@agent-of-mkmeral

File: strands-command/scripts/javascript/process-input.cjs (hunk 1 above)
Comment (agent-of-mkmeral) [reply strands-command/scripts/javascript/process-input.cjs]: Done in `fffde45`. Created `BETA_SYSTEM_PROMPT.md` based on:
- My own system prompt structure
- [Agent Guidelines](https://github.com/strands-agents/docs/blob/main/team/AGENT_GUIDELINES.md)
- Skills/sub-agent context

The system prompt covers: identity, guidelines (add value or stay silent, keep it short, prove don't opine), capabilities overview, behavioral rules, output format, and anti-patterns.

Loaded by both `process-input.cjs` (reads the file for `INPUT_SYSTEM_PROMPT`) and `beta_agent_runner.py` (`_load_system_prompt()` with fallback).

File: strands-command/scripts/python/beta_agent_runner.py (hunk 2 above)
Comment (agent-of-mkmeral) [reply strands-command/scripts/python/beta_agent_runner.py]: Done in `fffde45`. Added `programmatic_tool_caller.py` as a local copy from [strands-agents/tools#387](https://github.com/strands-agents/tools/pull/387).

The beta runner tries to load it in priority order:
1. `from strands_tools import programmatic_tool_caller` (when it merges into the tools package)
2. Local copy via `importlib.util` from `scripts/python/programmatic_tool_caller.py`
3. Graceful skip if neither available

Once tools#387 merges, the local copy becomes a no-op fallback.

File: strands-command/scripts/python/beta_agent_runner.py (hunk 3 above)
Comment (agent-of-mkmeral) [reply strands-command/scripts/python/beta_agent_runner.py:362]: Done in `fffde45`. The correct signature is `agent.tool.skills(skill_name="task-reviewer")` (checked the skills tool description 😄).

Implementation:
- `process-input.cjs` outputs `agent_mode` (e.g., `reviewer`, `implement`, `adversarial-test`) in the parsed JSON
- `action.yml` reads it and passes as `AGENT_MODE` env var
- `beta_agent_runner.py` has a `SKILL_MAP` that maps modes → skills and calls `_activate_skill_for_mode(agent, mode)` after agent creation

Mapping:
```
/strands beta review     → task-reviewer
/strands beta implement  → task-implementer
/strands beta refine     → task-refiner
/strands beta adversarial-test → task-adversarial-tester
/strands beta release-notes    → task-release-notes
/strands beta meta-reason      → task-meta-reasoner
/strands beta release-digest   → task-release-digest
```

File: strands-command/agent-skills/task-adversarial-tester/SKILL.md (hunk 4 above)
Comment (agent-of-mkmeral) [reply strands-command/agent-skills/task-adversarial-tester/SKILL.md:2]: Done in `fffde45`.

**Meta-reasoner:** Added `task-meta-reasoner/SKILL.md` with the full meta-reasoning gate SOP.

**SOPs as skills (no duplication):** Implemented runtime conversion in `_convert_sops_to_skills()`:
1. `action.yml` copies `agent-sops/` to the working directory alongside `agent-skills/`
2. At runtime, `beta_agent_runner.py` reads `.sop.md` files, adds YAML frontmatter (name, description, allowed-tools), and writes them as `SKILL.md` in the skills directory
3. Dedicated skills (like `task-adversarial-tester`) take precedence — if a `SKILL.md` already exists, the SOP is skipped
4. No source files are modified — it's a one-way write into the skills dir

So `task-implementer.sop.md` → `agent-skills/task-implementer/SKILL.md` at runtime, with proper frontmatter injected.

**Comment by agent-of-mkmeral** [comment]
Addressed all 4 review threads in [`fffde45`](https://github.com/mkmeral/devtools/commit/fffde45):

| Thread | Change | Files |
|---|---|---|
| 1. System prompt | Created `BETA_SYSTEM_PROMPT.md` based on agent guidelines, my own prompt, and skills context | `BETA_SYSTEM_PROMPT.md`, `process-input.cjs`, `beta_agent_runner.py` |
| 2. Programmatic tool caller | Local copy from [tools#387](https://github.com/strands-agents/tools/pull/387), auto-upgrades when merged | `programmatic_tool_caller.py`, `beta_agent_runner.py` |
| 3. Skill activation | `/strands beta review` → activates `task-reviewer` skill via `agent.tool.skills()` | `process-input.cjs`, `action.yml`, `beta_agent_runner.py` |
| 4. Meta-reasoner + SOPs as skills | New skill + runtime SOP→SKILL.md conversion (no file duplication) | `task-meta-reasoner/SKILL.md`, `action.yml`, `beta_agent_runner.py` |

### Diff summary

```
 BETA_SYSTEM_PROMPT.md          |  60 +++   (new — system prompt)
 task-meta-reasoner/SKILL.md    |  79 +++   (new — meta-reasoner skill)
 programmatic_tool_caller.py    | 316 +++++ (new — local PTC copy)
 action.yml                     |   6 +     (agent_mode + agent-sops copy)
 process-input.cjs              |  54 +-    (system prompt from file, agent_mode output, meta-reason routing)
 beta_agent_runner.py           | 245 +++-- (system prompt loading, PTC, skill activation, SOP→skill conversion)
```

6 files changed, 724 insertions, 36 deletions. `agent_runner.py` still **UNCHANGED**.
